src/ansi-c/c_nondet_symbol_factory.cpp

```diff
@@ -0,0 +1,403 @@
+    init_code.move_to_operands(input_code);
+  }
+  inputs.clear();
+}
+
+class symbol_factoryt
+{
+  code_blockt &init_code;
+  bool assume_non_null;
+  symbol_tablet &symbol_table;
+  input_recordert &input_recorder;
+  namespacet ns;
+
+public:
+  symbol_factoryt(
+    code_blockt &_init_code,
+    bool _assume_non_null,
+    symbol_tablet &_symbol_table,
+    input_recordert &input_recorder):
+      init_code(_init_code),
+      assume_non_null(_assume_non_null),
+      symbol_table(_symbol_table),
+      input_recorder(input_recorder),
+      ns(_symbol_table)
+  {}
+
+  exprt allocate_object(
+    const exprt &,
+    const typet &,
+    const source_locationt &,
+    const bool);
+
+  void gen_nondet_init(
+    const exprt &expr,
+    const source_locationt &loc);
+};
+
+/*******************************************************************\
+
+Function: symbol_factoryt::allocate_object
+
+ Inputs:
+
+ Outputs:
+
+ Purpose:
+
+\*******************************************************************/
+
+exprt symbol_factoryt::allocate_object(
+  const exprt &target_expr,
+  const typet &allocate_type,
+  const source_locationt &loc,
+  const bool static_lifetime)
+{
+  // Will generate the following code:
+  //   <allocate_type> tmp$1;
+  const typet &allocate_type_resolved=ns.follow(allocate_type);
+  const typet &target_type=ns.follow(target_expr.type().subtype());
+  bool cast_needed=allocate_type_resolved!=target_type;
+  symbolt &aux_symbol=
+    c_declare_new_tmp_symbol(init_code, symbol_table, loc, allocate_type);
+  aux_symbol.is_static_lifetime=static_lifetime;
+
+  exprt object=aux_symbol.symbol_expr();
+  exprt aoe=address_of_exprt(object);
+  if(cast_needed)
+  {
+    aoe=typecast_exprt(aoe, target_expr.type());
+  }
+  // Will generate the following code:
+  //   <target_expr> = &tmp$1
+  code_assignt assign(target_expr, aoe);
+  assign.add_source_location()=loc;
+  init_code.copy_to_operands(assign);
+
+  // Will generate the following code (stored for later):
+  //   INPUT("tmp$1", tmp$1);
+  input_recorder.record_input(aux_symbol.base_name, object);
+
+  return aoe;
+}
+
+/*******************************************************************\
+
+Function: symbol_factoryt::gen_nondet_init
+
+ Inputs:
+  expr -
+  loc -
+
+ Outputs:
+
+ Purpose:
+
+\*******************************************************************/
+
+void symbol_factoryt::gen_nondet_init(
+  const exprt &expr,
+  const source_locationt &loc)
+{
+  const typet &type=ns.follow(expr.type());
+
+  if(type.id()==ID_pointer)
+  {
+    // dereferenced type
+    const pointer_typet &pointer_type=to_pointer_type(type);
+    const typet &subtype=ns.follow(pointer_type.subtype());
+
+    code_labelt set_null_label;
+    code_labelt init_done_label;
+    std::string init_done_label_str;
+
+    if(!assume_non_null)
+    {
+      // Will generate the following code:
+      //   IF NONDET(_Bool) != FALSE THEN GOTO <set_null_label_str>
```

Comment:
Thanks a lot for the detailed comments. Could you please add the overall code in one place to understand what's intended to happen here at high level?